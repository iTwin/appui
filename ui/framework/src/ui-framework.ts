/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
export { default as UiFramework } from "./ui-framework/UiFramework";

export * from "./ui-framework/AppState";
export * from "./ui-framework/FrameworkState";
export * from "./ui-framework/UiFramework";
export * from "./ui-framework/CoreToolDefinitions";

export * from "./ui-framework/clientservices/IModelServices";
export * from "./ui-framework/clientservices/ProjectServices";

export * from "./ui-framework/feedback/ValidationTextbox";
export * from "./ui-framework/feedback/ElementTooltip";

export * from "./ui-framework/messages/InputField";
export * from "./ui-framework/messages/Pointer";

export * from "./ui-framework/oidc/SignIn";
export * from "./ui-framework/oidc/SignOut";

export * from "./ui-framework/openimodel/ApplicationHeader";
export * from "./ui-framework/openimodel/BlockingPrompt";
export * from "./ui-framework/openimodel/IModelCard";
export * from "./ui-framework/openimodel/IModelList";
export * from "./ui-framework/openimodel/IModelOpen";
export * from "./ui-framework/openimodel/IModelViewPicker";
export * from "./ui-framework/openimodel/Navigation";
export * from "./ui-framework/openimodel/ProjectDialog";
export * from "./ui-framework/openimodel/ProjectDropdown";
export * from "./ui-framework/openimodel/ProjectTabs";

export * from "./ui-framework/imodel-components/visibility-tree/VisibilityTree";

export * from "./ui-framework/overallcontent/OverallContent";
export * from "./ui-framework/overallcontent/state";

export * from "./ui-framework/pickers/ListPicker";
export * from "./ui-framework/pickers/ModelSelector";
export * from "./ui-framework/pickers/ViewSelector";

export * from "./ui-framework/messages/AppNotificationManager";
export * from "./ui-framework/configurableui/ConfigurableUiContent";
export * from "./ui-framework/configurableui/ConfigurableUiControl";
export * from "./ui-framework/configurableui/ConfigurableUiManager";
export * from "./ui-framework/content/ContentControl";
export * from "./ui-framework/content/ContentGroup";
export * from "./ui-framework/content/ContentLayout";
export * from "./ui-framework/content/ContentViewManager";
export * from "./ui-framework/dragdrop/DragDropLayerManager";
export * from "./ui-framework/dragdrop/ZoneTargets";
export * from "./ui-framework/frontstage/Frontstage";
export * from "./ui-framework/frontstage/FrontstageComposer";
export * from "./ui-framework/frontstage/FrontstageDef";
export * from "./ui-framework/frontstage/FrontstageManager";
export * from "./ui-framework/frontstage/FrontstageProvider";
export * from "./ui-framework/frontstage/NestedFrontstage";
export * from "./ui-framework/shared/IconComponent";
export * from "./ui-framework/shared/Item";
export * from "./ui-framework/shared/ItemDefBase";
export * from "./ui-framework/shared/ItemMap";
export * from "./ui-framework/shared/ItemProps";
export * from "./ui-framework/keyboardshortcut/KeyboardShortcut";
export * from "./ui-framework/keyboardshortcut/KeyboardShortcutMenu";
export * from "./ui-framework/messages/MessageManager";
export * from "./ui-framework/ModalDialogManager";
export * from "./ui-framework/frontstage/ModalFrontstage";
export * from "./ui-framework/navigationaids/NavigationAidControl";
export * from "./ui-framework/widgets/NavigationWidget";
export * from "./ui-framework/widgets/StackedWidget";
export * from "./ui-framework/messages/StandardMessageBox";
export * from "./ui-framework/widgets/StatusBar";
export * from "./ui-framework/widgets/StatusBarWidgetControl";
export * from "./ui-framework/zones/StatusBarZone";
export * from "./ui-framework/workflow/Task";
export * from "./ui-framework/widgets/ToolbarWidgetBase";
export * from "./ui-framework/widgets/ToolWidget";
export * from "./ui-framework/content/ViewportContentControl";
export * from "./ui-framework/widgets/Widget";
export * from "./ui-framework/widgets/WidgetControl";
export * from "./ui-framework/widgets/WidgetDef";
export * from "./ui-framework/widgets/WidgetFactory";
export * from "./ui-framework/workflow/Workflow";
export * from "./ui-framework/zones/FrameworkZone";
export * from "./ui-framework/zones/toolsettings/ToolSettingsZone";
export * from "./ui-framework/zones/toolsettings/ToolUiManager";
export * from "./ui-framework/zones/toolsettings/ToolUiProvider";
export * from "./ui-framework/zones/toolsettings/DefaultToolSettingsProvider";

export * from "./ui-framework/zones/Zone";
export * from "./ui-framework/zones/ZoneDef";

export * from "./ui-framework/toolbar/ActionItemButton";
export * from "./ui-framework/toolbar/GroupItem";
export * from "./ui-framework/toolbar/ToolButton";

export * from "./ui-framework/backstage/Backstage";
export * from "./ui-framework/backstage/FrontstageLaunch";
export * from "./ui-framework/backstage/CommandLaunch";
export * from "./ui-framework/backstage/TaskLaunch";

export * from "./ui-framework/navigationaids/CubeNavigationAid";
export * from "./ui-framework/navigationaids/SheetNavigationAid";
export * from "./ui-framework/navigationaids/SheetsModalFrontstage";
export * from "./ui-framework/navigationaids/StandardRotationNavigationAid";

export * from "./ui-framework/settings/Settings";

export * from "./ui-framework/statusfields/ActivityCenter";
export * from "./ui-framework/statusfields/MessageCenter";
export * from "./ui-framework/statusfields/SnapMode";
export * from "./ui-framework/statusfields/PromptField";
export * from "./ui-framework/statusfields/SelectionInfo";

export * from "./ui-framework/theme/ThemeManager";

export * from "./ui-framework/tools/AnalysisAnimation";
export * from "./ui-framework/tools/AnalysisAnimationToolSettings";
export * from "./ui-framework/tools/ScheduleAnimation";
export * from "./ui-framework/tools/ScheduleAnimationToolSettings";

export * from "./ui-framework/utils/ViewUtilities";
export * from "./ui-framework/utils/redux-ts";
export * from "./ui-framework/utils/PropsHelper";

export * from "./ui-framework/syncui/SyncUiEventDispatcher";
export * from "./ui-framework/syncui/BooleanListener";

// Set the version number so it can be found at runtime. BUILD_SEMVER is replaced at build time by the webpack DefinePlugin.
declare var BUILD_SEMVER: string;
if ((typeof (BUILD_SEMVER) !== "undefined") && (typeof window !== "undefined") && window) {
  if (!(window as any).iModelJsVersions)
    (window as any).iModelJsVersions = new Map<string, string>();
  (window as any).iModelJsVersions.set("ui-framework", BUILD_SEMVER);
}

/** @docs-package-description
 * The ui-framework package contains application fragments for Login, Project, iModel and View selection,
 * and configuration of the application UI with the Backstage, Frontstages, Widgets, etc.
 * For more information, see [learning about ui-framework]($docs/learning/framework/index.md).
 */
/**
 * @docs-group-description Backstage
 * Classes for working with a Backstage
 */
/**
 * @docs-group-description ClientServices
 * Classes for working with services for iModels, Login and Projects
 */
/**
 * @docs-group-description ConfigurableUi
 * Classes for working with the Application UI Configuration
 */
/**
 * @docs-group-description ContentView
 * Classes for working with a Content View, Group, Layout or Control
 */
/**
 * @docs-group-description Dialog
 * Classes for working with a modal dialog
 */
/**
 * @docs-group-description DragDrop
 * Classes for managing DragDrop API drag layers
 */
/**
 * @docs-group-description Frontstage
 * Classes for working with a Frontstage
 */
/**
 * @docs-group-description Item
 * Classes for working with an Item in a Tool Widget, Navigation Widget or Backstage
 */
/**
 * @docs-group-description KeyboardShortcut
 * Classes for working with Keyboard Shortcuts
 */
/**
 * @docs-group-description NavigationAids
 * Classes for working with Navigation Aids
 */
/**
 * @docs-group-description Notification
 * Classes for working with a Notification or Message
 */
/**
 * @docs-group-description OIDC
 * Classes for working with the OpenID Connect (OIDC) protocol
 */
/**
 * @docs-group-description OpenIModel
 * Classes for working with the UI for opening an iModel
 */
/**
 * @docs-group-description OverallContent
 * Classes for working with the overall content of the application UI
 */
/**
 * @docs-group-description Picker
 * Classes for working with various pickers
 */
/**
 * @docs-group-description Settings
 * Classes for Settings page
 */
/**
 * @docs-group-description State
 * Classes for maintaining state
 */
/**
 * @docs-group-description StatusBar
 * Classes for defining a StatusBar
 */
/**
 * @docs-group-description SyncUi
 * Classes for informing UI components to sync/refresh their display
 */

/**
 * @docs-group-description WorkflowTask
 * Classes for working a Workflow or Task
 */
/**
 * @docs-group-description Tools
 * Core Tool definitions
 */
/**
 * @docs-group-description ToolSettings
 * Classes for working Tool Settings
 */
/**
 * @docs-group-description Utilities
 * Various utility classes for working with a UI and Redux.
 */
/**
 * @docs-group-description Widget
 * Classes for working with a Widget
 */
/**
 * @docs-group-description Zone
 * Classes for working with a Zone
 */
