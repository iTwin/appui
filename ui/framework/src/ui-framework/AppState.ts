/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module AppState */

import { createAction, ActionsUnion } from "./utils/redux-ts";

// cSpell:ignore configurableui snapmode toolprompt
/** Action Ids used by redux and to send sync UI components. Typically used to refresh visibility or enable state of control.
 *  Since these are also used as sync ids they should be in lowercase.
 */
export const enum AppStateActionId {
  SetNumItemsSelected = "appstate:set-num-items-selected",
}

/** The portion of state managed by the AppStateReducer. */
export interface AppState {
  numItemsSelected: number;
}

/** used on first call of AppStateReducer */
const initialState: AppState = {
  /** number of selected items in Presentation Selection */
  numItemsSelected: 0,
};

/** An object with a function that creates each AppStateReducer that can be handled by our reducer. */ // tslint:disable-next-line:variable-name
export const AppStateActions = {
  setNumItemsSelected: (numSelected: number) => createAction(AppStateActionId.SetNumItemsSelected, numSelected),
};

/** Union of AppState Redux actions  */
export type AppStateActionsUnion = ActionsUnion<typeof AppStateActions>;

/** Handles actions to update AppState. */
export function AppStateReducer(state: AppState = initialState, _action: AppStateActionsUnion): AppState {
  switch (_action.type) {
    case AppStateActionId.SetNumItemsSelected: {
      if (undefined !== _action.payload)
        return { ...state, numItemsSelected: _action.payload };
      break;
    }

  }

  return state;
}
