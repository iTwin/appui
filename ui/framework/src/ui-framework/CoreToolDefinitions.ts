/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tools */

// cSpell:ignore configurableui
import {
  FitViewTool, FlyViewTool, IModelApp, PanViewTool, RotateViewTool, SelectionTool, ViewToggleCameraTool, WalkViewTool,
  WindowAreaTool, ZoomViewTool, ViewClipByPlaneTool,
  ViewClipDecorationProvider,
} from "@bentley/imodeljs-frontend";
import { ViewFlags } from "@bentley/imodeljs-common";
import { ToolItemDef } from "./shared/ToolItemDef";

/** Utility Class that provides definitions of tools provided by iModel.js core. These definitions can be used to populate the UI.
 * @public
 */
// istanbul ignore next
export class CoreTools {
  public static get fitViewCommand() {
    return new ToolItemDef({
      toolId: FitViewTool.toolId,
      iconSpec: "icon-fit-to-view",
      label: () => FitViewTool.flyover,
      tooltip: () => FitViewTool.description,
      execute: () => { IModelApp.tools.run(FitViewTool.toolId, IModelApp.viewManager.selectedView, true); },
    });
  }

  public static get windowAreaCommand() {
    return new ToolItemDef({
      toolId: WindowAreaTool.toolId,
      iconSpec: "icon-window-area",
      label: () => WindowAreaTool.flyover,
      tooltip: () => WindowAreaTool.description,
      execute: () => { IModelApp.tools.run(WindowAreaTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get zoomViewCommand() {
    return new ToolItemDef({
      toolId: ZoomViewTool.toolId,
      iconSpec: "icon-zoom",
      label: () => ZoomViewTool.flyover,
      tooltip: () => ZoomViewTool.description,
      execute: () => { IModelApp.tools.run(ZoomViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get panViewCommand() {
    return new ToolItemDef({
      toolId: PanViewTool.toolId,
      iconSpec: "icon-hand-2",
      label: () => PanViewTool.flyover,
      tooltip: () => PanViewTool.description,
      execute: () => { IModelApp.tools.run(PanViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get rotateViewCommand() {
    return new ToolItemDef({
      toolId: RotateViewTool.toolId,
      iconSpec: "icon-gyroscope",
      label: () => RotateViewTool.flyover,
      tooltip: () => RotateViewTool.description,
      execute: () => { IModelApp.tools.run(RotateViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get walkViewCommand() {
    return new ToolItemDef({
      toolId: WalkViewTool.toolId,
      iconSpec: "icon-walk",
      label: () => WalkViewTool.flyover,
      tooltip: () => WalkViewTool.description,
      execute: () => { IModelApp.tools.run(WalkViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get selectElementCommand() {
    return new ToolItemDef({
      toolId: SelectionTool.toolId,
      iconSpec: "icon-cursor",
      label: () => SelectionTool.flyover,
      tooltip: () => SelectionTool.description,
      execute: () => { IModelApp.tools.run(SelectionTool.toolId); },
    });
  }

  public static get toggleCameraViewCommand() {
    return new ToolItemDef({
      toolId: ViewToggleCameraTool.toolId,
      iconSpec: "icon-camera",
      label: () => ViewToggleCameraTool.flyover,
      tooltip: () => ViewToggleCameraTool.description,
      execute: () => { IModelApp.tools.run(ViewToggleCameraTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get flyViewCommand() {
    return new ToolItemDef({
      toolId: FlyViewTool.toolId,
      iconSpec: "icon-airplane",
      label: () => FlyViewTool.flyover,
      tooltip: () => FlyViewTool.description,
      execute: () => { IModelApp.tools.run(FlyViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  // note current ViewClipByPlaneTool is not automatically registered so the app must call ViewClipByPlaneTool.register();
  public static get sectionByPlaneCommand() {
    return new ToolItemDef({
      toolId: ViewClipByPlaneTool.toolId,
      iconSpec: "icon-plane",
      label: () => ViewClipByPlaneTool.flyover,
      tooltip: () => ViewClipByPlaneTool.description,
      execute: () => {
        const vp = IModelApp.viewManager.selectedView;
        if (!vp || !vp.view.is3d())
          return;

        // Turn on clip volume flag for section tools
        const viewFlags: ViewFlags = vp.view.viewFlags.clone();
        viewFlags.clipVolume = true;
        vp.viewFlags = viewFlags;

        IModelApp.tools.run(ViewClipByPlaneTool.toolId, ViewClipDecorationProvider.create());
      },
    });
  }

}
