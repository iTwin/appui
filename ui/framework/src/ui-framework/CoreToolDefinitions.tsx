/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tools */

import * as React from "react";

// cSpell:ignore configurableui keyinbrowser
import {
  FitViewTool, FlyViewTool, IModelApp, PanViewTool, RotateViewTool, SelectionTool, ViewToggleCameraTool, WalkViewTool,
  WindowAreaTool, ZoomViewTool, ViewClipByPlaneTool, ViewUndoTool, ViewRedoTool,
  ViewClipDecorationProvider,
} from "@bentley/imodeljs-frontend";
import { PopupButton, PopupButtonChildrenRenderPropArgs } from "./toolbar/PopupButton";
import { ViewFlags } from "@bentley/imodeljs-common";
import { ToolItemDef } from "./shared/ToolItemDef";
import { CustomItemDef } from "./shared/CustomItemDef";
import { KeyinBrowser } from "./keyinbrowser/KeyinBrowser";
import { SyncUiEventId } from "./syncui/SyncUiEventDispatcher";
import { BaseItemState } from "../ui-framework/shared/ItemDefBase";
import { ContentViewManager } from "./content/ContentViewManager";

/** Utility Class that provides definitions of tools provided by iModel.js core. These definitions can be used to populate the UI.
 * @public
 */
// istanbul ignore next
export class CoreTools {
  /** Get the CustomItemDef for PopupButton
   * @beta
   */
  public static get keyinBrowserButtonItemDef() {
    return new CustomItemDef({
      customId: "uif:keyinbrowser",
      reactElement: (
        <PopupButton iconSpec="icon-process" labelKey="UiFramework:keyinbrowser.label" betaBadge={true}>
          {this._renderKeyInBrowser}
        </PopupButton>
      ),
    });
  }

  private static _renderKeyInBrowser = ({ closePanel }: PopupButtonChildrenRenderPropArgs) => {
    return (
      <KeyinBrowser onExecute={closePanel} />
    );
  }

  public static get fitViewCommand() {
    return new ToolItemDef({
      toolId: FitViewTool.toolId,
      iconSpec: FitViewTool.iconSpec,
      label: () => FitViewTool.flyover,
      tooltip: () => FitViewTool.description,
      execute: () => { IModelApp.tools.run(FitViewTool.toolId, IModelApp.viewManager.selectedView, true); },
    });
  }

  public static get windowAreaCommand() {
    return new ToolItemDef({
      toolId: WindowAreaTool.toolId,
      iconSpec: WindowAreaTool.iconSpec,
      label: () => WindowAreaTool.flyover,
      tooltip: () => WindowAreaTool.description,
      execute: () => { IModelApp.tools.run(WindowAreaTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get zoomViewCommand() {
    return new ToolItemDef({
      toolId: ZoomViewTool.toolId,
      iconSpec: ZoomViewTool.iconSpec,
      label: () => ZoomViewTool.flyover,
      tooltip: () => ZoomViewTool.description,
      execute: () => { IModelApp.tools.run(ZoomViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get panViewCommand() {
    return new ToolItemDef({
      toolId: PanViewTool.toolId,
      iconSpec: PanViewTool.iconSpec,
      label: () => PanViewTool.flyover,
      tooltip: () => PanViewTool.description,
      execute: () => { IModelApp.tools.run(PanViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get rotateViewCommand() {
    return new ToolItemDef({
      toolId: RotateViewTool.toolId,
      iconSpec: RotateViewTool.iconSpec,
      label: () => RotateViewTool.flyover,
      tooltip: () => RotateViewTool.description,
      execute: () => { IModelApp.tools.run(RotateViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get walkViewCommand() {
    return new ToolItemDef({
      toolId: WalkViewTool.toolId,
      iconSpec: WalkViewTool.iconSpec,
      label: () => WalkViewTool.flyover,
      tooltip: () => WalkViewTool.description,
      execute: () => { IModelApp.tools.run(WalkViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get selectElementCommand() {
    return new ToolItemDef({
      toolId: SelectionTool.toolId,
      iconSpec: SelectionTool.iconSpec,
      label: () => SelectionTool.flyover,
      tooltip: () => SelectionTool.description,
      execute: () => {
        IModelApp.tools.run(SelectionTool.toolId);
      },
    });
  }

  public static get toggleCameraViewCommand() {
    return new ToolItemDef({
      toolId: ViewToggleCameraTool.toolId,
      iconSpec: ViewToggleCameraTool.iconSpec,
      label: () => ViewToggleCameraTool.flyover,
      tooltip: () => ViewToggleCameraTool.description,
      execute: () => { IModelApp.tools.run(ViewToggleCameraTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  public static get flyViewCommand() {
    return new ToolItemDef({
      toolId: FlyViewTool.toolId,
      iconSpec: FlyViewTool.iconSpec,
      label: () => FlyViewTool.flyover,
      tooltip: () => FlyViewTool.description,
      execute: () => { IModelApp.tools.run(FlyViewTool.toolId, IModelApp.viewManager.selectedView); },
    });
  }

  // note current ViewClipByPlaneTool is not automatically registered so the app must call ViewClipByPlaneTool.register();
  public static get sectionByPlaneCommand() {
    return new ToolItemDef({
      toolId: ViewClipByPlaneTool.toolId,
      iconSpec: ViewClipByPlaneTool.iconSpec,
      label: () => ViewClipByPlaneTool.flyover,
      tooltip: () => ViewClipByPlaneTool.description,
      execute: () => {
        const vp = IModelApp.viewManager.selectedView;
        if (!vp || !vp.view.is3d())
          return;

        // Turn on clip volume flag for section tools
        const viewFlags: ViewFlags = vp.view.viewFlags.clone();
        viewFlags.clipVolume = true;
        vp.viewFlags = viewFlags;

        IModelApp.tools.run(ViewClipByPlaneTool.toolId, ViewClipDecorationProvider.create());
      },
    });
  }

  // TODO - Need to provide a sync message that is fire when the Undo/Redo button needs to be refreshed in the
  // active view.
  public static get viewUndoCommand() {
    return new ToolItemDef({
      toolId: ViewUndoTool.toolId,
      iconSpec: ViewUndoTool.iconSpec,
      label: () => ViewUndoTool.flyover,
      tooltip: () => ViewUndoTool.description,
      execute: () => { IModelApp.tools.run(ViewUndoTool.toolId, IModelApp.viewManager.selectedView); },
      stateSyncIds: [SyncUiEventId.ActiveContentChanged],
      stateFunc: (currentState: Readonly<BaseItemState>): BaseItemState => {
        const returnState: BaseItemState = { ...currentState };
        const activeContentControl = ContentViewManager.getActiveContentControl();
        if (activeContentControl && activeContentControl.viewport)
          returnState.isEnabled = activeContentControl.viewport.isRedoPossible;
        return returnState;
      },
    });
  }

  public static get viewRedoCommand() {
    return new ToolItemDef({
      toolId: ViewRedoTool.toolId,
      iconSpec: ViewRedoTool.iconSpec,
      label: () => ViewRedoTool.flyover,
      tooltip: () => ViewRedoTool.description,
      execute: () => { IModelApp.tools.run(ViewRedoTool.toolId, IModelApp.viewManager.selectedView); },
      stateSyncIds: [SyncUiEventId.ActiveContentChanged],
      stateFunc: (currentState: Readonly<BaseItemState>): BaseItemState => {
        const returnState: BaseItemState = { ...currentState };
        const activeContentControl = ContentViewManager.getActiveContentControl();
        if (activeContentControl && activeContentControl.viewport)
          returnState.isEnabled = activeContentControl.viewport.isRedoPossible;
        return returnState;
      },
    });
  }
}
