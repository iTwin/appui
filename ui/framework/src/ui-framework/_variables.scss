/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
@import "@bentley/ui-core/lib/ui-core/index";

$uifw-popup-border-color:         $buic-foreground-disabled;
$uifw-popup-border-width:         1px;
$uifw-popup-border-radius:        3px;
$uifw-popup-content-padding:      5px;
$uifw-popup-title-bar-min-height: 24px;
$uifw-popup-opacity:              0.80;
