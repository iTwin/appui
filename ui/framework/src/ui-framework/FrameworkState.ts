/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module FrameworkState */

import { OverallContentState, OverallContentReducer } from "./overallcontent/state";
import { ConfigurableUiState, ConfigurableUiReducer } from "./configurableui/state";
import { AppState, AppStateReducer } from "./AppState";
import { combineReducers } from "./utils/redux-ts";

/** Interface combining all the state Framework state interfaces. */
export interface FrameworkState {
  overallContentState: OverallContentState;
  configurableUiState: ConfigurableUiState;
  appState: AppState;
}

/** Framework reducer that combines the [[OverallContentReducer]],  [[ConfigurableUiReducer]] and OidcReducer. */
export const FrameworkReducer = combineReducers({   // tslint:disable-line:variable-name
  overallContentState: OverallContentReducer,
  configurableUiState: ConfigurableUiReducer,
  appState: AppStateReducer,
});
