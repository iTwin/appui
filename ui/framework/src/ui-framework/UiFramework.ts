/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Utilities */
import { OidcFrontendClientConfiguration } from "@bentley/imodeljs-clients";
import { I18N } from "@bentley/imodeljs-i18n";
import { ActivityLoggingContext, Guid } from "@bentley/bentleyjs-core";
import { ProjectServices } from "./clientservices/ProjectServices";
import { DefaultProjectServices } from "./clientservices/DefaultProjectServices";
import { IModelServices } from "./clientservices/IModelServices";
import { DefaultIModelServices } from "./clientservices/DefaultIModelServices";
import { Store } from "redux";
import { OidcClientWrapper } from "@bentley/imodeljs-frontend";
import { AnalysisAnimationTool } from "./tools/AnalysisAnimation";
import { SyncUiEventDispatcher } from "./syncui/SyncUiEventDispatcher";
import { FrameworkState } from "./FrameworkState";

/**
 * Manages the Redux store, I18N service and iModel, Project and Login services for the ui-framework package.
 */
export class UiFramework {
  private static _projectServices?: ProjectServices;
  private static _iModelServices?: IModelServices;
  private static _i18n?: I18N;
  private static _store?: Store<any>;
  private static _complaint = "UiFramework not initialized";
  private static _frameworkStateKeyInStore: string = "frameworkState";  // default name

  /**
   * Called by IModelApp to initialize the UiFramework
   * @param store The single redux store created by the IModelApp.
   * @param i18n The internationalization service created by the IModelApp.
   * @param oidcConfig Optional configuration for authenticating user.
   * @param frameworkStateKey The name of the key used by the IModelApp when adding the UiFramework state into the Redux store. If not defined "frameworkState" is assumed.
   * @param projectServices Optional IModelApp defined projectServices.If not specified DefaultProjectServices will be used.
   * @param iModelServices Optional IModelApp defined iModelServices.If not specified DefaultIModelServices will be used.
   */
  public static async initialize(store: Store<any>, i18n: I18N, oidcConfig?: OidcFrontendClientConfiguration, frameworkStateKey?: string, projectServices?: ProjectServices, iModelServices?: IModelServices) {
    UiFramework._store = store;
    UiFramework._i18n = i18n;
    if (frameworkStateKey)
      UiFramework._frameworkStateKeyInStore = frameworkStateKey;

    const frameworkNamespace = UiFramework._i18n.registerNamespace("UiFramework");
    const readFinishedPromise = frameworkNamespace.readFinished;

    // register UiFramework provided tools
    AnalysisAnimationTool.register(frameworkNamespace);

    UiFramework._projectServices = projectServices ? projectServices : new DefaultProjectServices();
    UiFramework._iModelServices = iModelServices ? iModelServices : new DefaultIModelServices();

    if (oidcConfig) {
      const initOidcPromise = OidcClientWrapper.initialize(new ActivityLoggingContext(Guid.createValue()), oidcConfig);
      return Promise.all([readFinishedPromise, initOidcPromise]);
    }
    return readFinishedPromise;
  }

  public static terminate() {
    UiFramework._store = undefined;
    UiFramework._frameworkStateKeyInStore = "frameworkState";

    if (UiFramework._i18n)
      UiFramework._i18n.unregisterNamespace("UiFramework");
    UiFramework._i18n = undefined;
    UiFramework._projectServices = undefined;
    UiFramework._iModelServices = undefined;
  }

  public static get frameworkStateKey(): string {
    return UiFramework._frameworkStateKeyInStore;
  }

  public static get frameworkState(): FrameworkState | undefined {
    // tslint:disable-next-line:no-string-literal
    return UiFramework.store.getState()[UiFramework.frameworkStateKey];
  }

  public static get store(): Store<any> {
    if (!UiFramework._store)
      throw new Error(UiFramework._complaint);
    return UiFramework._store;
  }

  public static get i18n(): I18N {
    if (!UiFramework._i18n)
      throw new Error(UiFramework._complaint);
    return UiFramework._i18n;
  }

  public static get projectServices(): ProjectServices {
    if (!UiFramework._projectServices)
      throw new Error(UiFramework._complaint);
    return UiFramework._projectServices!;
  }

  public static get iModelServices(): IModelServices {
    if (!UiFramework._iModelServices)
      throw new Error(UiFramework._complaint);
    return UiFramework._iModelServices!;
  }

  public static dispatchActionToStore(type: string, payload: any, immediateSync = false) {
    UiFramework.store.dispatch({ type, payload });
    if (immediateSync)
      SyncUiEventDispatcher.dispatchImmediateSyncUiEvent(type);
    else
      SyncUiEventDispatcher.dispatchSyncUiEvent(type);
  }
}

export default UiFramework;
