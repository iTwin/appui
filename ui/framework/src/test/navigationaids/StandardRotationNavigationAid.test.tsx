/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import * as React from "react";
import { mount, shallow } from "enzyme";
import { expect } from "chai";

import { StandardRotationNavigationAid } from "../../ui-framework";
import TestUtils from "../TestUtils";

// NEEDSWORK_MODULARIZATION needs work after modularization reorganization. Might just need snap update, might now be wrong.
describe("StandardRotationNavigationAid", () => {

  before(async () => {
    await TestUtils.initializeUiFramework();
  });

  describe("<StandardRotationNavigationAid />", () => {

    it("should render", () => {
      mount(<StandardRotationNavigationAid />);
    });

    it("renders correctly", () => {
      shallow(<StandardRotationNavigationAid />).should.matchSnapshot();
    });

    it("should expand on click and change on item click", () => {
      const wrapper = mount(<StandardRotationNavigationAid />);

      expect(wrapper.find("span.icon-cube-faces-top").length).to.eq(1);
      expect(wrapper.find("div.nz-toolbar-item-expandable-group-tool-item").length).to.eq(0);
      wrapper.find("div.icon-button").simulate("click");

      expect(wrapper.find("div.nz-toolbar-item-expandable-group-tool-item").length).to.be.greaterThan(0);
      wrapper.find("div.nz-toolbar-item-expandable-group-tool-item").at(1).simulate("click");
      expect(wrapper.find("span.icon-cube-faces-top").length).to.eq(0);
      expect(wrapper.find("span.icon-cube-faces-bottom").length).to.eq(1);

      wrapper.unmount();
    });

  });

});
