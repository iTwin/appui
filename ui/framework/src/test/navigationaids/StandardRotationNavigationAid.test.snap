// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandardRotationNavigationAid <StandardRotationNavigationAid /> renders correctly 1`] = `
<div
  className="uifw-standard-rotation-navigation"
>
  <ExpandableButton
    button={
      <ToolbarIcon
        className="icon-button"
        icon={
          <span
            className="three-d-icon icon icon-cube-faces-top"
          />
        }
        onClick={[Function]}
      />
    }
    className="expandable"
    direction={3}
  />
</div>
`;
